### PR TITLE
CLOUD-229 Upgrade pxc test 

### PR DESCRIPTION
**Pull request by cap1984**

### CODE REVIEW

File: e2e-tests/upgrade/run [hunk 1]
```diff
@@ -0,0 +1,68 @@
+#!/bin/bash
```
Comment (delgod) [inline e2e-tests/upgrade/run]: add test to `e2e-tests/run`

File: e2e-tests/upgrade/run [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+
+set -o errexit
+set -o xtrace
+
+test_dir=$(realpath $(dirname $0))
+. ${test_dir}/../functions
+
+CLUSTER='upgrade'
+IMAGE='percona/percona-xtradb-cluster-operator:1.0.0'
```
Comment (delgod) [inline e2e-tests/upgrade/run]: cache images
```
+IMAGE_TO_UPDATE=$IMAGE
+IMAGE_PXC_TO_UPDATE=$IMAGE_PXC
+IMAGE_PROXY_TO_UPDATE=$IMAGE_PROXY
+IMAGE_BACKUP_TO_UPDATE=$IMAGE_BACKUP
 IMAGE='percona/percona-xtradb-cluster-operator:1.0.0'
 IMAGE_PXC='percona/percona-xtradb-cluster-operator:1.0.0-pxc'
 IMAGE_PROXY='percona/percona-xtradb-cluster-operator:1.0.0-proxysql'
 IMAGE_BACKUP='percona/percona-xtradb-cluster-operator:1.0.0-backup'
-IMAGE_TO_UPDATE='perconalab/percona-xtradb-cluster-operator:master'
```

File: e2e-tests/upgrade/run [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+    kubectl rollout status deployment/percona-xtradb-cluster-operator
+
+    wait_pod "$(kubectl get pods --selector=name=percona-xtradb-cluster-operator \
+                                -o custom-columns='NAME:.metadata.name,IMAGE:.spec.containers[0].image' \
+                                | grep "${IMAGE_TO_UPDATE}" |  awk '{print $1}')"
+
+    sleep 20
+    check_pxc_livenes "${CLUSTER}" 3
+
+    kubectl apply -f "${test_dir}/conf/${CLUSTER}-master.yml"
```
Comment (delgod) [inline e2e-tests/upgrade/run]: use cached images
```
-    kubectl apply -f "${test_dir}/conf/${CLUSTER}-master.yml"
+    IMAGE_PXC=$IMAGE_PXC_TO_UPDATE
+    IMAGE_PROXY=$IMAGE_PROXY_TO_UPDATE
+    IMAGE_BACKUP=$IMAGE_BACKUP_TO_UPDATE
+    apply_config "${test_dir}/conf/${CLUSTER}.yml"
```

File: e2e-tests/upgrade/run [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+        compare_mysql_cmd "select-1" "SELECT * from myApp.myApp;" "-h ${cluster_name}-pxc-$i.${cluster_name}-pxc -uroot -proot_password"
+    done
+}
+
+main() {
+    create_namespace $namespace
+    deploy_operator
+
+    spinup_pxc "${CLUSTER}" "${test_dir}/conf/${CLUSTER}.yml"
+
```
Comment (delgod) [inline e2e-tests/upgrade/run]: check statefulset generation and operator image here

File: e2e-tests/upgrade/run [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+        -p'{"spec":{"template":{"spec":{"containers":[{"name":"percona-xtradb-cluster-operator","image":"'"${IMAGE_TO_UPDATE}"'"}]}}}}'
+    kubectl rollout status deployment/percona-xtradb-cluster-operator
+
+    wait_pod "$(kubectl get pods --selector=name=percona-xtradb-cluster-operator \
+                                -o custom-columns='NAME:.metadata.name,IMAGE:.spec.containers[0].image' \
+                                | grep "${IMAGE_TO_UPDATE}" |  awk '{print $1}')"
+
+    sleep 20
+    check_pxc_livenes "${CLUSTER}" 3
+
```
Comment (delgod) [inline e2e-tests/upgrade/run]: check statefulset generation and operator image and cluster image here

File: e2e-tests/upgrade/run [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+    kubectl apply -f "${test_dir}/conf/${CLUSTER}-master.yml"
+    check_pxc_livenes "${CLUSTER}" 3
+
```
Comment (delgod) [inline e2e-tests/upgrade/run]: check statefulset generation and cluster image here

File: e2e-tests/upgrade/run (hunk 2 above)
Comment (cap1984) [reply e2e-tests/upgrade/run]: done

File: e2e-tests/upgrade/run (hunk 6 above)
Comment (cap1984) [reply e2e-tests/upgrade/run]: done

File: e2e-tests/upgrade/run (hunk 5 above)
Comment (cap1984) [reply e2e-tests/upgrade/run]: done

File: e2e-tests/upgrade/run (hunk 4 above)
Comment (cap1984) [reply e2e-tests/upgrade/run]: done

File: e2e-tests/upgrade/run (hunk 3 above)
Comment (delgod) [reply e2e-tests/upgrade/run]: could you please use patch command ^^^

File: e2e-tests/upgrade/run [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+    wait_cluster_consistency "${cluster_name}" "${cluster_size}"
+    
+    wait_for_running "${cluster_name}-pxc" "${cluster_size}"
+
+    for i in $(seq 0 $(($cluster_size-1))); do
+        compare_mysql_cmd "select-1" "SELECT * from myApp.myApp;" "-h ${cluster_name}-pxc-$i.${cluster_name}-pxc -uroot -proot_password"
+    done
+}
+
+compare_generation() {
```
Comment (delgod) [inline e2e-tests/upgrade/run]: it is possible to pass the expected number as the second argument of the function, in such case creation of files is not needed

File: e2e-tests/upgrade/run [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+    spinup_pxc "${CLUSTER}" "${test_dir}/conf/${CLUSTER}.yml"
+
+    compare_generation "before_upgrade"
+
+    kubectl patch deployment "${OPERATOR_NAME}" \
+        -p'{"spec":{"template":{"spec":{"containers":[{"name":"'"${OPERATOR_NAME}"'","image":"'"${IMAGE_TO_UPDATE}"'"}]}}}}'
+    kubectl rollout status deployment/"${OPERATOR_NAME}"
+
+    echo 'Waiting for operator upgrade'
+    until [[ $(kubectl get pods --selector=name=${OPERATOR_NAME} \
```
Comment (delgod) [inline e2e-tests/upgrade/run]:
```suggestion
    wait_pod $(get_operator_pod)
```

File: e2e-tests/upgrade/run [hunk 9]
```diff
@@ -0,0 +1,119 @@
...
+    compare_generation "cluster_upgrade"
+
+    check_pxc_livenes "${CLUSTER}" 3
+
+    if [[ "${IMAGE_PROXY_TO_UPDATE}" == $(kubectl get pxc "${CLUSTER}" -o jsonpath='{.spec.proxysql.image}') \
+          && "${IMAGE_PXC_TO_UPDATE}" == $(kubectl get pxc "${CLUSTER}" -o jsonpath='{.spec.pxc.image}') ]]; then
+        : Cluster image has been updated correctly
+    else
+        echo 'Cluster image has not been updated'
+        destroy "${namespace}"
```
Comment (delgod) [inline e2e-tests/upgrade/run]: it is hard to debug if test destroyed namespace

File: e2e-tests/upgrade/run [hunk 10]
```diff
@@ -0,0 +1,119 @@
...
+        exit 1
+    fi
```
Comment (delgod) [inline e2e-tests/upgrade/run]: it is also needed to check if mysql_upgrade was run
```
for i in $(seq 0 $(($cluster_size-1))); do
    kubectl logs ${cluster_name}-pxc-$i \
        | grep mysql_upgrade
done
```

File: e2e-tests/upgrade/run (hunk 9 above)
Comment (cap1984) [reply e2e-tests/upgrade/run]: removed

File: e2e-tests/upgrade/run (hunk 7 above)
Comment (cap1984) [reply e2e-tests/upgrade/run]: done

File: e2e-tests/upgrade/run (hunk 8 above)
Comment (cap1984) [reply e2e-tests/upgrade/run]: we need to exclude the pod in terminating status on the image check line. Either we wait here or  we have figure out something else

**Pull request merged**